pkg/derivation/kv.go

```diff
@@ -0,0 +1,95 @@
+package derivation
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"os"
+)
+
+type KVStore interface {
+	Get(key string) ([]byte, error)
+	Set(key string, value []byte) error
+}
+
+// File system KV
+
+type FSKVStore struct{}
+
+func (kv *FSKVStore) Get(key string) ([]byte, error) {
+	return os.ReadFile(key)
+}
+
+func (kv *FSKVStore) Set(key string, value []byte) error {
+	return fmt.Errorf("Writing not supported")
+}
+
+// Input substitution calculations
+
+type InputSubstKV struct {
+	memo  map[string][]byte // Memoized values
+	store KVStore           // Underlying store
+}
+
+func NewInputSubstKV(store KVStore) *InputSubstKV {
+	return &InputSubstKV{
+		memo:  make(map[string][]byte),
+		store: store,
+	}
+}
+
+func (kv *InputSubstKV) Get(key string) ([]byte, error) {
+	value, ok := kv.memo[key]
+
+	if !ok {
```

Comment:
I guess we can just do an early return if okay, and then indent the whole rest one level less.